File: outputs/src/@opencode/sdk/examples/basic.ts
/**
 * @opencode/sdk - Basic Usage Example
 *
 * This example demonstrates basic usage of the OpenCode TypeScript SDK.
 */

import {
  OpenCodeClient,
  SdkError,
  ToolCall,
} from '../src/index';

/**
 * Example: Basic session creation and tool execution.
 */
async function basicExample(): Promise<void> {
  console.log('=== OpenCode SDK Basic Example ===\n');

  // Create client with API key
  const client = new OpenCodeClient({
    baseUrl: process.env.OPENCODE_BASE_URL ?? 'http://localhost:8080/api',
    auth: { key: process.env.OPENCODE_API_KEY ?? 'sk-your-api-key' },
    timeout: 30000,
  });

  try {
    // Create a new session
    console.log('1. Creating session...');
    const session = await client.createSession('Hello, OpenCode!');
    console.log(`   Created session: ${session.sessionId}`);
    console.log(`   Status: ${session.status}`);
    console.log(`   Message count: ${session.messageCount}\n`);

    // Add a message
    console.log('2. Adding message...');
    const addResult = await client.addMessage(
      session.sessionId,
      'Can you read the file at /tmp/test.txt?',
      'user'
    );
    console.log(`   Session: ${addResult.sessionId}`);
    console.log(`   Total messages: ${addResult.messageCount}\n`);

    // List sessions
    console.log('3. Listing sessions...');
    const sessions = await client.listSessions(10, 0);
    console.log(`   Found ${sessions.length} sessions`);
    sessions.forEach((s) => {
      console.log(`   - ${s.id}: ${s.messageCount} messages`);
    });
    console.log();

    // List available tools
    console.log('4. Listing available tools...');
    const tools = await client.listTools();
    console.log(`   Found ${tools.length} tools`);
    tools.slice(0, 5).forEach((t) => {
      console.log(`   - ${t.name}: ${t.description.slice(0, 50)}...`);
    });
    if (tools.length > 5) {
      console.log(`   ... and ${tools.length - 5} more`);
    }
    console.log();

    // Execute a tool
    console.log('5. Executing tool (read file)...');
    const toolCall: ToolCall = {
      name: 'read',
      arguments: {
        path: '/tmp/test.txt',
      },
    };

    try {
      const result = await client.executeTool(toolCall);
      if (result.success) {
        console.log(`   Success! Result: ${result.result?.slice(0, 100)}...`);
      } else {
        console.log(`   Failed: ${result.error}`);
      }
    } catch (error) {
      if (error instanceof SdkError) {
        console.log(`   SDK Error [${error.code}]: ${error.message}`);
      } else {
        throw error;
      }
    }
    console.log();

    // Fork session
    console.log('6. Forking session...');
    const fork = await client.forkSession(session.sessionId, 1);
    console.log(`   Forked session: ${fork.id}`);
    console.log(`   Parent: ${fork.parentSessionId}`);
    console.log(`   Messages: ${fork.messageCount}\n`);

    console.log('=== Example completed successfully! ===');
  } catch (error) {
    if (error instanceof SdkError) {
      console.error(`SDK Error [${error.code}] (${error.category}): ${error.message}`);
      process.exit(1);
    }
    throw error;
  }
}

/**
 * Example: Local session (offline mode).
 */
async function localSessionExample(): Promise<void> {
  console.log('\n=== Local Session Example ===\n');

  const client = OpenCodeClient.builder()
    .apiKey('sk-local-key')
    .build();

  // Create local session (no server required)
  console.log('1. Creating local session...');
  await client.createLocalSession('Working offline');
  const localSession = await client.getLocalSession();
  console.log(`   Local session ID: ${localSession?.id ?? 'none'}`);
  console.log(`   Messages: ${localSession?.messageCount ?? 0}\n`);

  // Add messages locally
  console.log('2. Adding local messages...');
  await client.addLocalMessage('assistant', 'Hello! How can I help you?');
  await client.addLocalMessage('user', 'Just testing local mode.');
  const updated = await client.getLocalSession();
  console.log(`   Messages: ${updated?.messageCount ?? 0}\n`);

  console.log('=== Local session example completed! ===');
}

// Run examples
basicExample().catch(console.error);
localSessionExample().catch(console.error);

File: opencode-rust/crates/sdk/examples/basic.rs
//! Basic example of using the OpenCode SDK.
//!
//! This example demonstrates:
//! - Creating a client
//! - Creating a session
//! - Adding messages
//! - Executing tools
//! - Listing sessions

use opencode_sdk::tools::ToolCall;
use opencode_sdk::OpenCodeClient;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    eprintln!("OpenCode SDK Basic Example");
    eprintln!("Note: This example requires a running OpenCode server");

    // Create the client
    println!("Creating OpenCode client...");
    let client = OpenCodeClient::builder()
        .base_url("http://localhost:8080/api")
        .api_key("sk-your-api-key-here")
        .timeout(std::time::Duration::from_secs(30))
        .build()?;

    println!("Client created successfully!");
    println!("  Base URL: {}", client.config().base_url);

    // Create a new session with an initial prompt
    println!("\nCreating a new session...");
    let session = client
        .create_session(Some("Hello, OpenCode! Can you help me with a task?"))
        .await?;

    println!("Session created!");
    println!("  Session ID: {}", session.session_id);
    println!("  Status: {}", session.status);
    println!("  Message count: {}", session.message_count);

    // Add a follow-up message
    println!("\nAdding a follow-up message...");
    let response = client
        .add_message(
            &session.session_id.to_string(),
            Some("user"),
            "Can you list the files in the current directory?",
        )
        .await?;

    println!("Message added!");
    println!("  New message count: {}", response.message_count);

    // List all sessions
    println!("\nListing all sessions...");
    let sessions = client.list_sessions(Some(10), Some(0)).await?;

    println!("Found {} sessions:", sessions.len());
    for s in &sessions {
        println!("  - {} ({} messages)", s.id, s.message_count);
    }

    // List available tools
    println!("\nListing available tools...");
    let tools = client.list_tools().await?;

    println!("Found {} tools:", tools.len());
    for tool in &tools {
        println!("  - {}: {}", tool.name, tool.description);
    }

    // Execute a tool (if the server supports it)
    println!("\nAttempting to execute 'read' tool...");
    let tool_call = ToolCall::new("read", serde_json::json!({ "file_path": "Cargo.toml" }));

    match client.execute_tool(tool_call).await {
        Ok(result) => {
            println!("Tool executed successfully!");
            println!("  Success: {}", result.is_success());
            if result.is_success() {
                let result_str = result.result.as_deref().unwrap_or("");
                println!("  Result: {}", &result_str[..result_str.len().min(200)]);
            }
        }
        Err(e) => {
            println!("Tool execution failed: {}", e);
            println!("  (This is expected if the server is not running or tool is not available)");
        }
    }

    // Fork the session (create a branch)
    println!("\nForking the session at message index 0...");
    match client
        .fork_session(&session.session_id.to_string(), 0)
        .await
    {
        Ok(forked) => {
            println!("Session forked!");
            println!("  New Session ID: {}", forked.id);
            println!("  Parent Session ID: {:?}", forked.parent_session_id);
        }
        Err(e) => {
            println!("Fork failed: {}", e);
        }
    }

    // Abort the original session
    println!("\nAborting the original session...");
    match client.abort_session(&session.session_id.to_string()).await {
        Ok(()) => {
            println!("Session aborted successfully!");
        }
        Err(e) => {
            println!("Abort failed: {}", e);
        }
    }

    println!("\nDone!");

    Ok(())
}
